### PR TITLE
Css 10676/add http proxy

### PR DESCRIPTION
**Pull request by SimoneDutto**
## Description

This pr implements the http proxy methods.

> fly-by moved utils functions to a more appropiate file

Fixes _JIRA/GitHub issue number_

## Engineering checklist
*Check only items that apply*

- [ ] Documentation updated
- [ ] Covered by unit tests
- [ ] Covered by integration tests

### CODE REVIEW

File: internal/rpc/httpproxy_test.go [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+	"github.com/canonical/jimm/v3/internal/rpc"
+)
+
+func TestProxyHTTP(t *testing.T) {
+	c := qt.New(t)
+	ctx := context.Background()
+	// we expect the controller to respond with TLS
+	fakeController := httptest.NewTLSServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		_, err := w.Write([]byte("OK"))
+		c.Assert(err, qt.IsNil)
```
Comment (kian99) [inline internal/rpc/httpproxy_test.go]: I'd like to see one test where the controller doesn't return an HTTP 200 and the client gets that back.

**Review by kian99 — DISMISSED** [review]
LGTM with a request for extra test

File: internal/rpc/httpproxy.go [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+	"net/http"
+	"net/url"
+
+	"github.com/juju/zaputil"
+	"github.com/juju/zaputil/zapctx"
+
+	"github.com/canonical/jimm/v3/internal/dbmodel"
+)
+
+type HTTPClientOptions struct {
```
Comment (alesstimec) [inline internal/rpc/httpproxy.go]: godoc, please

File: internal/rpc/httpproxy_test.go (hunk 1 above)
Comment (SimoneDutto) [reply internal/rpc/httpproxy_test.go]: added

**Review by kian99 — DISMISSED** [review]

File: internal/rpc/httpproxy.go [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+	TLSConfig *tls.Config
```
Comment (alesstimec) [inline internal/rpc/httpproxy.go:21]: we could pass in an HTTP client instead?

File: internal/rpc/httpproxy.go (hunk 2 above)
Comment (alesstimec) [reply internal/rpc/httpproxy.go]: also these seem more like `HTTPOptions` not just `client` options: e.g. the URL is a property of the request. and i don't think this type even needs to be exported

File: internal/rpc/httpproxy.go [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+	http.Error(w, "Gateway timeout", http.StatusGatewayTimeout)
+}
+
+func doRequest(ctx context.Context, w http.ResponseWriter, req *http.Request, opt HTTPClientOptions) error {
+	client := &http.Client{
+		Transport: &http.Transport{
+			TLSClientConfig: opt.TLSConfig,
+		},
+	}
+	req.RequestURI = ""
```
Comment (alesstimec) [inline internal/rpc/httpproxy.go:75]: suggestion: also do `req = req.Clone()`

File: internal/rpc/httpproxy_test.go [hunk 5]
```diff
@@ -0,0 +1,112 @@
+// Copyright 2024 Canonical.
+package rpc_test
```
Comment (alesstimec) [inline internal/rpc/httpproxy_test.go:3]: new line between copyright and package, please

**Review by alesstimec — DISMISSED** [review]
LGTM modulo a few comments

File: internal/rpc/httpproxy.go (hunk 3 above)
Comment (SimoneDutto) [reply internal/rpc/httpproxy.go:21]: tls config must be used when creatinig the struct. I don't like the idea of creating the httpclient outside of the doRequest function, however i don't really have a strong opinion on that.
Why do you think it's better?

**Review by kian99 — APPROVED** [review]

**Review by ale8k — APPROVED** [review]

File: internal/rpc/httpproxy.go (hunk 3 above)
Comment (ale8k) [reply internal/rpc/httpproxy.go:21]: Generally speaking you'd have clients specific for different purposes with different roundtrippers

**Pull request merged**